### ISSUE TITLE
MapProperty: adding entry from an empty provider should not make the map "unset"

### ISSUE THREAD

**Issue by tkrullmann** [gradle/gradle#13364]
When calling `put` on a `MapProperty` with a `Provider` that evaluates to missing, the whole `MapProperty` will evaluate to missing.

### Expected Behavior

The `MapProperty` should include all the entries and skip the entry that has no value.

### Current Behavior

The whole `MapProperty` will have no value.

### Context

Trying to launch an external process using the worker API and `ExecOperations`, where several optional environment variables are collected from properties, and the environment is passed to a `WorkParameters` instance in a `MapProperty`. If any of the optional properties is not set, the whole map becomes unset.

### Steps to Reproduce 

Example build.gradle.kts:

```kotlin
val map = objects.mapProperty<String, String>()
map.put("key1", "value1")
map.put("key2", provider { null })
println(map.orNull)
```

Expected output:
```
{key1=value1}
```

Actual output:
```
null
```

**Comment by tkrullmann** [comment]
The problem seems to be in `DefaultMapProperty.PlusCollector`:

https://github.com/gradle/gradle/blob/dec0179942960245ce9ce728bccf765b1c51080a/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/DefaultMapProperty.java#L535-L541

If `right` evaluates to missing, then that "missing" result is returned, even if entries were already added to `dest`.

**Comment by lacasseio** [comment]
I agree this is an important usability issue of the `MapProperty`. At the very least, Gradle should mention which entry had no value. Sometimes there can be multiple entries coming from a wide variety of sources. A single message there was no value is confusing. Despite that, I believe the no value entry should just be skipped as it doesn't make sense to have "one bad apple" to dictate the no value of the provider.

The workaround is to use `putAll` but that adds a lot of boilerplate code and is tricky to get the right implicit task dependencies.

**Comment by Vampire** [comment]
Still relevant, please reopen

**Comment by Vampire** [comment]
Still relevant in 7.4.2, please reopen

**Comment by CRogers** [comment]
I just ran into this too - very confusing!

**Comment by cloudshiftchris** [comment]
ugh.  Confusing indeed - those contracts are broken.  This was good for an hour or so of troubleshooting.

**Comment by gilday** [comment]
Reproduced in 8.0.2 with this reproducer

```kotlin
abstract class PrintTask : DefaultTask() {

    @get:Input
    abstract val map: MapProperty<String, String>

    @TaskAction
    fun print() {
        logger.info("map value {}", map.get())
    }
}

tasks.register("print", PrintTask::class) {
    map.put("FOO", providers.provider { "foo" })

    // this unexpectedly clears all previously set and future values of all keys
    map.put("BAR", providers.provider { null })

    map.put("BAZ", providers.provider { "baz" })
}
```
